src/android/UploadTask.java

```diff
@@ -0,0 +1,326 @@
+                        .putBoolean(KEY_OUTPUT_IS_ERROR, true)
+                        .putString(KEY_OUTPUT_FAILURE_REASON, "User cancelled")
+                        .putBoolean(KEY_OUTPUT_FAILURE_CANCELED, true)
+                        .build()
+                );
+            } else {
+                // But if it was not it must be a connectivity problem or
+                // something similar so we retry later
+                Log.e(TAG, "doWork: Call failed, retrying later", e);
+                return Result.retry();
+            }
+        }
+
+        // Start building the output data
+        final Data.Builder outputData = new Data.Builder()
+                .putString(KEY_OUTPUT_ID, id)
+                .putBoolean(KEY_OUTPUT_IS_ERROR, false)
+                .putInt(KEY_OUTPUT_STATUS_CODE, response.code());
+
+        // Try read the response body, if any
+        try {
+            final String res = response.body() != null ? response.body().string() : "";
+            final String filename = "upload-response-" + getId() + ".cached-response";
+
+            try (FileOutputStream fos = getApplicationContext().openFileOutput(filename, Context.MODE_PRIVATE)) {
+                fos.write(res.getBytes(StandardCharsets.UTF_8));
+            }
+
+            outputData.putString(KEY_OUTPUT_RESPONSE_FILE, filename);
+
+        } catch (IOException e) {
+            // Should never happen, but if it does it has something to do with reading the response
+            Log.e(TAG, "doWork: Error while reading the response body", e);
+
+            // But recover and replace the body with something else
+            outputData.putString(KEY_OUTPUT_RESPONSE_FILE, null);
+        }
+
+        return Result.success(outputData.build());
+    }
+
+    /**
+     * Called internally by the custom request body provider each time 8kio are written.
+     */
+    private void handleProgress(long bytesWritten, long totalBytes) {
+        if (isStopped()) {
+            currentCall.cancel();
+            return;
+        }
+
+        double percent = ((double) bytesWritten / (double) totalBytes) * 100.0;
+        Log.i(TAG, "handleProgress: " + getId() + " Progress: " + (int) percent);
+
+        setProgressAsync(new Data.Builder()
+                .putString(KEY_PROGRESS_ID, getInputData().getString(KEY_INPUT_ID))
+                .putInt(KEY_PROGRESS_PERCENT, (int) percent)
+                .build()
+        );
+        setForegroundAsync(createForegroundInfo((int) percent));
+    }
+
+    /**
+     * Create the OkHttp request that will be used, already filled with input data.
+     *
+     * @return A ready to use OkHttp request
+     * @throws FileNotFoundException If the file to upload can't be found
+     */
+    @NonNull
+    private Request createRequest() throws FileNotFoundException {
+        final String filepath = getInputData().getString(KEY_INPUT_FILEPATH);
+        assert filepath != null;
```

Comment:
We should ensure these exceptions are caught and they're not crashing the app. Plugin crashes crash the whole app using it. 🙄

Reply:
It is handled above in doWork() and will count as a failed upload with an appropriate message.